File: flex+bison/eval.c
#include <stdio.h>
#include <stdlib.h>
#include "eval.h"

int evalEXP(EXP *e)
{
	switch (e->kind) {
		case k_expressionKindIdentifier:
			fprintf(stderr, "Error: (line %d) Cannot evaluate the value of an identifier\n", e->lineno);
			exit(1);
		case k_expressionKindIntLiteral:
			return e->val.intLiteral;
		case k_expressionKindAddition:
			return evalEXP(e->val.binary.lhs) + evalEXP(e->val.binary.rhs);
		case k_expressionKindSubtraction:
			return evalEXP(e->val.binary.lhs) - evalEXP(e->val.binary.rhs);
		case k_expressionKindMultiplication:
			return evalEXP(e->val.binary.lhs) * evalEXP(e->val.binary.rhs);
		case k_expressionKindDivision:
			return evalEXP(e->val.binary.lhs) / evalEXP(e->val.binary.rhs);
		default: 
			fprintf(stderr, "Error: (line %d) Unknown kind for an expression node\n", e->lineno);
			exit(1);
	}
}

File: ocamllex+menhir/eval.ml
open Tree

let rec eval_exp (e: exp) =
  match e with
  | Identifier (_, line) -> failwith ("(line " ^ string_of_int line ^ ") Cannot evaluate the value of an identifier\n")
  | IntLiteral (v, _) -> v
  | Binary (lhs, rhs, op) -> eval_binary_exp lhs rhs op
and eval_binary_exp (lhs: exp) (rhs: exp) (op: operator) =
  match op with
  | Plus -> eval_exp lhs + eval_exp rhs
  | Minus -> eval_exp lhs - eval_exp rhs
  | Mult -> eval_exp lhs * eval_exp rhs
  | Div -> eval_exp lhs / eval_exp rhs
;;

let eval_program (prog: program) =
  match prog with
  | Program e -> eval_exp e
;;